### PR TITLE
Entry API

### LINKED ISSUE #810: Entry API
### Would you like to work on this feature?

yes

### What problem are you trying to solve?

A higher-level API than #806 for addressing the get-or-create resource use-case.

### Describe the solution you'd like

See the discussion in #806.

But basically, `let cm: ConfigMap = (cms: Api<ConfigMap>).entry("foo").await?.or_create(|| generate_cm()).await?;`.

### Describe alternatives you've considered

Only expose `Api::try_get`, and let users handle get-or-create themselves.

### Documentation, Adoption, Migration Strategy

_No response_

### Target crate for feature

kube-client

### PR DESCRIPTION
**Pull request by teozkr**
## Motivation

Fixes #810

## Solution

Introduces a new method `Api::entry`, analogous to std's `HashMap::entry`, allowing for a (hopefully) ergonomic API for get-and-create and get-and-modify use cases.

The API should be similar to `HashMap::entry`, but adds an explicit `Entry::sync()` method for writing back any changes to K8s, allowing the user to control how and when batching happens.

### CODE REVIEW

File: kube-client/src/api/entry.rs
Comment (teozkr) [inline kube-client/src/api/entry.rs:113]: Not sure about whether we actually want to do this, or just do a NOOP here.

It feels a bit footgunny that we "give up" the optimistic lock based on a runtime condition...

File: kube-client/src/api/entry.rs
Comment (teozkr) [inline kube-client/src/api/entry.rs:146]: Maybe we should "just" implement `Debug` for `Api`?

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs:113]: On the flip side, if we do implement `get' for `Clean` then we might also want to do the same for `VacantEntry`?

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:111]: I guess this is going to be the most contentious part as using `::replace` here makes it a bit brittle w.r.t. interacting field managers / whatever admission controllers is battling for the object's rights, while everything in kubernetes is moving to a patch/apply world.

Kind of wondering if it would be useful for json-patches (or apply configurations down the line in #649) to be constructible in some other way from the root structs, and then they could be buffered herein for a bigger patch call. At least that could help control the size / sanity of the api calls so they are less likely to fail / require retries.

Comment (clux) [reply kube-client/src/api/entry.rs:111]: Do you know what client-go does in this situation? I think they have some implicit syncing going on.

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:131]: this helpful insert needs to check for a `generateName` before deciding to potentially blanket override it

**Review by clux — COMMENTED** [review]
very clean in general, left one minor nit, but also a more future oriented problem with using replace

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs:111]: We can't use `apply` or `patch` here, because the get-or-create and get-and-modify patterns are both inherently prone to race conditions, and patches bypass the `resourceVersion` check. `apply` is even worse, since for get-and-modify you end up taking ownership of the whole object.

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs:131]: Yeah, we should probably reject objects with `generateName` set entirely, since that doesn't really make sense for `get-and-create`. Maybe we should even reject objects where `obj.meta.name` != `self.name` as well (when set manually).

File: kube-client/src/api/entry.rs
Comment (clux) [reply kube-client/src/api/entry.rs:111]: Yeah, you are right. I think there's probably two avenues we can take for these helpers; ones that modify objects (using `replace`) that maybe could be retry-wrapped somehow, and ones that consist of combinable sets of json patches (using `patch`).

Incidentally, there was another instance of the first category here propping up in https://github.com/kube-rs/kube-rs/issues/428#issuecomment-1033592469 .

Comment (teozkr) [reply kube-client/src/api/entry.rs:111]: I'd rather separate them into "things that require (fallible) server coordination" and "things that manipulate client-side objects".

Yes, users will have to retry when this fails but I think we have to leave this up to them rather than doing this ourselves. Sometimes the `default` function may not be safe to retry, sometimes they have a coarser-grained retry policy that they'd want to bubble up to (such as when running under the `Controller`/`applier`).

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs]: Added warnings in 03a2967, is that fine or do we want to promote them to full-blown errors?

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs:167]: To me, this is the big remaining issue before I'd be happy to merge this.

Comment (teozkr) [reply kube-client/src/api/entry.rs:167]: The more I think about this, the more I lean towards NOOP. That way, we have a clean flow of "`sync` fails if there are changes, and anyone else has modified the object since the call to `Api::entry`"

Comment (clux) [reply kube-client/src/api/entry.rs]: yeah, I agree. Sync on Clean state should probably be a noop. If people do conditional things to the entry, and then always sync, then we know whether or not it needs to be a noop. I would probably add a one-liner about it for the `sync` fn though.

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:18]: I feel a `no_run` doc example here would be helpful for users!

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:63]: could/should also have smaller `no_run` doc examples for these other main interfaces (or_insert, get_mut, sync)

`Api::entry` is the main thing that needs one though

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:50]: Since we are making `Entry` enum public, we should probably have language more like BTreeMap's `Entry` in these main methods on `Entry`. e.g. this one, maybe restructure to:

```suggestion
    /// Provides in-place mutable access to an occupied entry
    ///
    /// The entry is occupied if it exists on the API, or is queued for creation using [`Entry::or_insert`])
```

to have a more entry-language like header, and then we can explain what that means outside the header.

File: kube-client/src/api/entry.rs
Comment (clux) [reply kube-client/src/api/entry.rs]: i'm inclined to error on misuse of name/namespace/generate_name properties because the whole abstraction is geared toward giving a name then mutating it / inserting a new one in its place. It feels pretty strange to suddenly send in an entry with a different name, or in a namespace the api was not configured for.

but otoh, we would never catch all the cases where the user subverts any validation we put in place (via `get_mut` say).. so maybe the helpful warnings are enough? or maybe it's better to put the errors and the `.sync` stage and always store the `name` property on all `Entry` types

**Review by clux — COMMENTED** [review]
This is looking very professional, had a good go over now and could only find a few nits on documentation. The integration tests are very good (i see the coverage has bounced back from the earlier pr state :+1:), and presumably these are pretty fast since it's just configmap mutations.

some unknowns on the potential error cases where people manually set metadata.{name,namespace,generate_name} but there's already a thread for it

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs]: Done.

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs]: I added an example to `Api::entry` that also covers `or_insert`, `and_modify`, and `sync`.

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs]: Done. Do we want to rename this to `save`?

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs:81]: Can we expect users to have read the enum-level docs already? I agree that the current wording is clunky, but I'd like to keep the emphasis on the unsaved/saved distinction (since this is the big difference compared to the `hash_map::Entry` API that the user is likely already familiar with).

**Review by clux — DISMISSED** [review]
had a quick look at the generated docs, looks great!

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs]: Yeah.. this is is a tricky aspect to balance. Making `or_insert` fallible would be annoying ergonomically for what is a programming error in 99% of cases, but I'm also not sure I'm comfortable turning it into a `panic!`.

Moving the check into `sync` does make some sense...

File: kube-client/src/api/entry.rs
Comment (clux) [reply kube-client/src/api/entry.rs]: either is fine with me. slightly leaning towards `commit` because that at least gives the correct impression that we are not fetching upstream changes :thinking:

Comment (teozkr) [reply kube-client/src/api/entry.rs]: Renamed to `commit`.

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:45]: straggler from rename ^

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:475]: you could put a pub method on `CommitError` if you want an easier way to determine this maybe? :thinking:

File: kube-client/src/api/entry.rs
Comment (clux) [inline kube-client/src/api/entry.rs:282]: these errors look great. however, they do use the new 1.58 f-strings so will need to do the msrv bump: https://kube-rs.github.io/website/rust-version/#minimum-supported-rust-version

Comment (teozkr) [reply kube-client/src/api/entry.rs:282]: `thiserror` has had specific support for this syntax since before Rust mainlined it. For example, I still use rustc 1.56.1 as my primary local compiler (due to #746), and it builds fine there.

File: kube-client/src/api/entry.rs
Comment (teozkr) [reply kube-client/src/api/entry.rs:475]: I'd rather not. Any actual users should just fail and retry at a level where it makes sense.

**Review by clux — APPROVED** [review]
looks very clean now, no other comments on this.
merge away at your leisure, and please open up any follow-up enhancement issue ideas (like the filthyref idea)

File: kube-client/src/api/entry.rs
Comment (clux) [reply kube-client/src/api/entry.rs]: it's much better now :+1:

**Pull request merged**
